### PR TITLE
Text APIs: interface re-design with name changes

### PR DESCRIPTION
**Pull request by astonzhang**
## Description ##
Text APIs: interface name change

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Interface revision, tests, (and when applicable, API doc)

## Comments ##
- Can we move it to mxnet.text asap?

### CODE REVIEW

File: docs/api/python/text/text.md [hunk 1]
```diff
@@ -41,7 +41,7 @@ fastText word embedding object by specifying the embedding name `fasttext` and
 the pre-trained file `wiki.simple.vec`.
 
 ```python
->>> fasttext_simple = text.embedding.TokenEmbedding.create('fasttext',
+>>> fasttext_simple = text.embedding._TokenEmbedding.create('fasttext',
```
Comment (szha) [inline docs/api/python/text/text.md]: `text.embedding.create`

File: docs/api/python/text/text.md [hunk 2]
```diff
@@ -121,13 +121,13 @@ by specifying their indices (2 and 1) and the `glossary.idx_to_vec` in
...
 
-## Glossary
+## CompositeEmbedding
 
 The glossary provides indexing and embedding for text tokens in a glossary. For
 each indexed token in a glossary, an embedding vector will be associated with
 it. Such embedding vectors can be loaded from externally hosted or custom
 pre-trained token embedding files, such as via instances of
-[`TokenEmbedding`](#mxnet.contrib.text.embedding.TokenEmbedding). 
+[`_TokenEmbedding`](#mxnet.contrib.text.embedding.TokenEmbedding).
```
Comment (szha) [inline docs/api/python/text/text.md]: This needs some work. There shouldn't be any reference to a private class.

**Comment by szha** [comment]
Pinging @piiswrong, @zhreshold.

**Comment by szha** [comment]
If we could agree on the interface in this PR, I'd suggest we consider moving this to `mxnet.text`.

File: docs/api/python/text/text.md [hunk 3]
```diff
@@ -137,14 +137,14 @@ candidate indices may be obtained via
 .. autosummary::
     :nosignatures:
 
-    Glossary
+    CompositeEmbedding
 ```
 
 To get all the valid names for pre-trained embeddings and files, we can use
-[`TokenEmbedding.get_embedding_and_pretrained_file_names`](#mxnet.contrib.text.embedding.TokenEmbedding.get_embedding_and_pretrained_file_names).
+[`_TokenEmbedding.get_embedding_and_pretrained_file_names`](#mxnet.contrib.text.embedding.TokenEmbedding.get_embedding_and_pretrained_file_names).
```
Comment (zhreshold) [inline docs/api/python/text/text.md]: `get_embedding_and_pretrained_file_names` function name too long and tedious IMO.

File: docs/api/python/text/text.md (hunk 2 above)
Comment (eric-haibin-lin) [reply docs/api/python/text/text.md]: Do you also need to update the reference  `mxnet.contrib.text.embedding.TokenEmbedding `?

File: docs/api/python/text/text.md (hunk 3 above)
Comment (astonzhang) [reply docs/api/python/text/text.md]: Thanks. Although it is possible to shorten it with some alias, I think this name is clearer.

File: python/mxnet/contrib/text/__init__.py [hunk 4]
```diff
@@ -19,6 +19,5 @@
 """This module includes utilities for indexing and embedding text."""
 
 from . import utils
-from . import indexer
+from . import vocab
```
Comment (szha) [inline python/mxnet/contrib/text/__init__.py]: Since `Vocabulary` is the only class that's in that namespace, we can consider removing this namespace by importing its members.

**Comment by szha** [comment]
Thanks for the fast turnaround, @astonzhang. Let's collectively address the naming and the interface definition first.

Requesting feedback from @zhreshold @piiswrong.

File: python/mxnet/contrib/text/__init__.py (hunk 4 above)
Comment (astonzhang) [reply python/mxnet/contrib/text/__init__.py]: This is intended for potential future extension.

Comment (szha) [reply python/mxnet/contrib/text/__init__.py]: Do you have such cases in mind? `text.vocab.Vocabulary` looks less clean than `text.Vocabulary`, and `vocab` is neither a proper word nor common enough to become an expected package name.

File: docs/api/python/text/text.md (hunk 3 above)
Comment (szha) [reply docs/api/python/text/text.md]: I'm inclined to what @zhreshold says. This is a lot of typing indeed... Consider a more concise name.

File: python/mxnet/contrib/text/__init__.py (hunk 4 above)
Comment (astonzhang) [reply python/mxnet/contrib/text/__init__.py]: Thanks. For instance, SubwordVocabulary may be added in the future.

File: docs/api/python/text/text.md (hunk 3 above)
Comment (astonzhang) [reply docs/api/python/text/text.md]: Since get_embedding_and_pretrained_file_names() takes embedding_name=None as its argument, how about get_pretrained_file_names() ?

Comment (szha) [reply docs/api/python/text/text.md]: Sounds good to me

**Comment by szha** [comment]
Hosted the doc at http://mxnet-doc.s3-accelerate.dualstack.amazonaws.com/api/python/text/text.html

File: docs/api/python/text/text.md [hunk 5]
```diff
@@ -120,73 +112,137 @@ by specifying their indices (2 and 1) and the `glossary.idx_to_vec` in
...
-The input counter whose keys are
-candidate indices may be obtained via
+The vocabulary builds indices for text tokens. Such indexed tokens can be used by token embedding
+instances. The input counter whose keys are candidate indices may be obtained via
 [`count_tokens_from_str`](#mxnet.contrib.text.utils.count_tokens_from_str).
 
+
 ```eval_rst
-.. currentmodule:: mxnet.contrib.text.glossary
+.. currentmodule:: mxnet.contrib.text.indexer
```
Comment (szha) [inline docs/api/python/text/text.md]: vocab

File: docs/api/python/text/text.md [hunk 6]
```diff
@@ -120,73 +112,137 @@ by specifying their indices (2 and 1) and the `glossary.idx_to_vec` in
...
+.. autosummary::
+    :nosignatures:
+
+    register
+    create
+    get_embedding_and_pretrained_file_names
+    GloVe
+    FastText
+    CustomEmbedding
+    CompositeEmbedding
```
Comment (szha) [inline docs/api/python/text/text.md]: Seems that this isn't in the right module

File: docs/api/python/text/text.md (hunk 3 above)
Comment (astonzhang) [reply docs/api/python/text/text.md]: Resolved.

File: docs/api/python/text/text.md (hunk 5 above)
Comment (astonzhang) [reply docs/api/python/text/text.md]: Resolved.

File: docs/api/python/text/text.md (hunk 6 above)
Comment (astonzhang) [reply docs/api/python/text/text.md]: Resolved.

File: docs/api/python/text/text.md [hunk 7]
```diff
@@ -120,73 +112,137 @@ by specifying their indices (2 and 1) and the `glossary.idx_to_vec` in
...
+
+
+## Text token embedding
+
+To load token embeddings from an externally hosted pre-trained token embedding file, such as those
+of GloVe and FastText, use
+[`_embedding.create(embedding_name, pretrained_file_name)`](#mxnet.contrib.text.embedding.create).
+
+To get all the available `embedding_name` and `pretrained_file_name`, use
+[`_embedding.get_pretrained_file_names()`](#mxnet.contrib.text.embedding.get_pretrained_file_names).
```
Comment (szha) [inline docs/api/python/text/text.md]: remove underscores. next time consider using global replacement with manual confirmation.

File: python/mxnet/contrib/text/__init__.py (hunk 4 above)
Comment (szha) [reply python/mxnet/contrib/text/__init__.py]: I think this only depends on tokenization. The class of vocabulary itself doesn't need any change.

File: docs/api/python/text/text.md (hunk 7 above)
Comment (astonzhang) [reply docs/api/python/text/text.md]: Thanks! Resolved.

**Comment by piiswrong** [comment]
Please move doc to contrib. Otherwise LGTM

File: docs/api/python/contrib/contrib.md [hunk 8]
```diff
@@ -0,0 +1,12 @@
+# Contrib APIs
```
Comment (szha) [inline docs/api/python/contrib/contrib.md]: Let's call this "Contrib Package".

File: docs/api/python/index.md [hunk 9]
```diff
@@ -151,3 +142,13 @@ imported by running:
 
    rtc/rtc.md
 ```
+
+## Contrib API
```
Comment (szha) [inline docs/api/python/index.md]: Same here.

**Comment by szha** [comment]
This PR looks good to me now. Since all review comments have been addressed, I'm merging this once CI passes. Thank @piiswrong @zhreshold for the timely reviews.
